scrapy_zyte_api/responses.py

```diff
@@ -166,10 +171,67 @@ def from_api_response(cls, api_response: Dict, *, request: Request = None):
 _API_RESPONSE = Dict[str, _JSON]
 
 
+def _check_response_size_limits(
+    expected_size: int,
+    warnsize: Optional[int],
+    maxsize: Optional[int],
+    request_url: str,
+) -> bool:
+    if warnsize and expected_size > warnsize:
+        logger.warning(
+            f"Expected response size {expected_size} larger than "
+            f"download warn size {warnsize} in request {request_url}."
+        )
+
+    if maxsize and expected_size > maxsize:
+        logger.warning(
+            f"Cancelling download of {request_url}: expected response size "
+            f"{expected_size} larger than download max size {maxsize}."
+        )
+        return False
+    return True
+
+
+def _response_max_size_exceeded(
+    api_response: _API_RESPONSE,
+    request: Request,
+    default_maxsize: Optional[int],
+    default_warnsize: Optional[int],
+) -> bool:
+    maxsize = request.meta.get("download_maxsize", default_maxsize)
+    warnsize = request.meta.get("download_warnsize", default_warnsize)
+
+    if "browserHtml" in api_response:
+        expected_size = len(api_response["browserHtml"].encode(_DEFAULT_ENCODING))
```

Comment:
Here while trying to limit the memory using DOWNLOAD_MAXSIZE we might create an additional memory spike, because we create another duplicate of browserHtml in memory, temporarily. 

It seems we need to ensure _response_max_size_exceeded never makes copies of large data from the response.

Reply:
I think you can either use sys.getsizeof (and maybe subtract some fixed overhead Python Unicode objects have), or consider the length of the unicode object instead of the length of the actual data, it could be good enough as well (though worse). Maybe there is some other solution.

Reply:
And, we already calculate base64 in two places
https://github.com/scrapy-plugins/scrapy-zyte-api/blob/main/scrapy_zyte_api/responses.py#L197
and 
https://github.com/scrapy-plugins/scrapy-zyte-api/blob/main/scrapy_zyte_api/responses.py#L127
But not sure if this is an issue and we need to fix it the PR.

Reply:
Another approach is to calculate the size and decoded/encoded version of the response body here before calling `from_api_response` and send the prepared body to `from_api_response`. In this case we make this expensive calculation only once.

Reply:
I was also thinking about moving it down the stack - check the size of the API response received by the client library before the json decoding. But it could make the library less compatible with Scrapy. Let's say you have an existing spider, which uses some download limit. You switch to scrapy-zyte-api for downloads, and maybe also enable data extraction. But the API response is larger than that raw response size. So, the limit becomes more aggressive, and you might drop some pages which were working before.

Because of this, the approach you've taken originally - checking httpResponseBody size and browserHtml size, ignoring everything else (e.g. structured data sizes, or screenshot sizes) makes sense to me.

Reply:
Ok, let me prepare another version with implementing check here https://github.com/scrapy-plugins/scrapy-zyte-api/blob/main/scrapy_zyte_api/responses.py#L197
and
https://github.com/scrapy-plugins/scrapy-zyte-api/blob/main/scrapy_zyte_api/responses.py#L127

Reply:
@kmike a new version is here.